builder/layer.go

```diff
@@ -0,0 +1,272 @@
+		log.Error("error starting 9p filesystem", "err", err)
+		return err
+	}
+	defer srcFS.Close()
+	outFS, err := add9pfs(cmd, outDir, "/out", true)
+	if err != nil {
+		log.Error("error starting 9p filesystem", "err", err)
+		return err
+	}
+	defer outFS.Close()
+
+	// run the job using the local flynn-host
+	cmd.Host = cluster.NewHost("build", "192.0.2.100:1113", nil, nil)
+
+	// prefix output with the name and script
+	logR, logW := io.Pipe()
+	defer logR.Close()
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		log.Error("error creating stdout pipe", "err", err)
+		return err
+	}
+	go io.Copy(logW, stdout)
+	stderr, err := cmd.StderrPipe()
+	if err != nil {
+		log.Error("error creating stderr pipe", "err", err)
+		return err
+	}
+	go io.Copy(logW, stderr)
+	go func() {
+		s := bufio.NewScanner(logR)
+		for s.Scan() {
+			fmt.Fprintf(
+				os.Stderr,
+				"%s: %s:%s: %s\n",
+				time.Now().Format("15:04:05.999"),
+				name, filepath.Base(script), s.Text(),
+			)
+		}
+	}()
+
+	// give the job some common permissions for installing software
+	cmd.LinuxCapabilities = append(host.DefaultCapabilities, "CAP_AUDIT_WRITE")
+
+	// run the job
+	log.Info("running build job", "name", name, "script", script, "hash", hash)
+	if err := cmd.Run(); err != nil {
+		log.Error("error running the build job", "err", err)
+		return err
+	}
+
+	// copy the layer to the store
+	f, err := os.Open(filepath.Join(outDir, "layer.squashfs"))
+	if err != nil {
+		log.Error("error opening squashfs layer", "err", err)
+		return err
+	}
+	defer f.Close()
+	w, err := store.LayerWriter()
+	if err != nil {
+		log.Error("error creating store layer writer", "err", err)
+		return err
+	}
+	if _, err := io.Copy(w, f); err != nil {
+		log.Error("error copying the layer to the store", "err", err)
+		w.Cancel()
+		return err
+	}
+	meta := map[string]string{
+		"flynn-build.name":   name,
+		"flynn-build.script": script,
+	}
+	layer, err = w.Commit(hash, meta)
+	if err != nil {
+		log.Error("error committing the layer to the store", "err", err)
+		return err
+	}
+
+	// write the layer to STDOUT
+	return json.NewEncoder(os.Stdout).Encode(layer)
+}
+
+// loadArtifact loads an artifact from ROOT/image/${name}.json, finding it
+// relative to the current binary's path ROOT/builder/bin/flynn-build
+func loadArtifact(name string) (*ct.Artifact, error) {
+	selfPath, err := filepath.Abs(os.Args[0])
+	if err != nil {
+		return nil, err
+	}
+	binDir := filepath.Dir(selfPath)
+
+	f, err := os.Open(filepath.Join(binDir, "..", "..", "image", name+".json"))
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+
+	var artifact ct.Artifact
+	return &artifact, json.NewDecoder(f).Decode(&artifact)
+}
+
+func generateHash(name string, inputs ...string) (string, error) {
+	h := sha512.New512_256()
+	h.Write([]byte(name))
+	for _, input := range inputs {
+		f, err := os.Open(input)
+		if err != nil {
+			return "", err
+		}
+		defer f.Close()
+		if _, err := io.Copy(h, f); err != nil {
```

Comment:
The name of the input and size should be written in the hash, to ensure that collisions can't happen if a piece of a file is removed from the end of one and added to the beginning of the next. This will probably not happen in practice, but it doesn't hurt to be correct.